File: Protocol/Protocol/Protocol/Engine/Engine.m
//
//  Engine.m
//  Protocol
//
//  Created by 김기현 on 2021/02/16.
//

#import <Foundation/Foundation.h>
#import "Engine.h"

@implementation Engine

- (id) copyWithZone: (NSZone *) zone {
    Engine * engineCopy;
    engineCopy = [[[self class] allocWithZone: zone] init];
    
    return engineCopy;
}

- (NSString *) description {
    return @"I am an engine. Vroom!";
} // description

@end // Engine

File: CarProtocol Swift/CarProtocolSwift/CarProtocolSwift/Engine/Engine.swift
//
//  Engine.swift
//  CarProtocolSwift
//
//  Created by 김기현 on 2021/02/17.
//

import Foundation

class Engine: NSCopying {
    required init() {
        
    }
    
    func copy(with zone: NSZone? = nil) -> Any {
        let engineCopy: Engine?
        engineCopy = type(of: self).init()
        
        guard let result = engineCopy else { fatalError("engine copy fail") }
        return result
    }
    
    func description() -> String {
        return "I am an engine. Vroom!"
    }
}
